### ISSUE TITLE
Add support for captions to the Gallery shortcode

### ISSUE THREAD

**Issue by kienstra** [ampproject/amp-wp#3658]
## Feature description

The `[gallery]` shortcode does not display captions. It'd be good to add support for them, as the Classic editor uses this for galleries.

A follow-up to #2855

---------------

## Acceptance criteria

1. If a `[gallery]` shortcode has images with captions, the captions should display in the `<amp-carousel>`.
2. Images without captions should still look good.

## Implementation brief

* PR #3285 [added logic](https://github.com/ampproject/amp-wp/pull/3285/files#diff-3f3a86f42d679f81a48116543bcb3b38R166) for creating an `<amp-carousel>` with slides that can have captions.

*  It'd be good to reuse as much of this as possible, though getting the actual caption text will be different for Gallery shortcodes.

## QA testing instructions

1. In `/wp-admin`, click 'Media' 
2. Click any image:
<img width="804" alt="media-here-image" src="https://user-images.githubusercontent.com/4063887/67738060-e9a71380-f9d2-11e9-964a-97f199ba46bf.png">

3. Add a caption, and make note of the ID:
<img width="956" alt="add-caption" src="https://user-images.githubusercontent.com/4063887/67738108-21ae5680-f9d3-11e9-8f85-c0df6330fe49.png">

4. Repeat steps 2-3 
5. You should now have at least 2 IDs for images
6. Create a new post in the block editor
7. Add a 'Shortcode' block
8. Add a Gallery shortcode, with IDs of the images from step 2. For example, if the IDs were `4572` and `3619`, the shortcode would be:

```html
[gallery ids="4572, 3619"]
```
9. Click 'Preview,' and go to the AMP URL
10. Expected: The captions should be visible in the gallery
11. Repeat steps 1-9, with galleries that have no images with captions, very long captions, only a few images with captions, etc...
12. Please re-test the [Gallery block testing steps](https://github.com/ampproject/amp-wp/pull/3285#issuecomment-546171299). PR #3659 refactors that code, but there's no change intended to the Gallery block.

## Demo

* 

## Changelog entry

*

**Comment by csossi** [comment]
Verified in QA:
![image](https://user-images.githubusercontent.com/16945664/70380844-79fc4080-190f-11ea-946c-ecd9521dd30e.png)

**Comment by csossi** [comment]
@kienstra , I retested as per step 12 and opted to "display as carousel" but AMP version of page isn't appearing as carousel:
![image](https://user-images.githubusercontent.com/16945664/70380941-d14ee080-1910-11ea-84bc-25aec59c1529.png)

**Comment by kienstra** [comment]
Hi @csossi,
Thanks a lot for testing this.

I'll make a note to look at this again, though it might be a few days.

**Comment by kienstra** [comment]
Hi @csossi,
Thanks for testing the Gallery block.

Maybe this wasn't the case before, but on the testing site, validation looks to be failing, and it's not actually an AMP URL:

https://story-test-wordpress-amp.pantheonsite.io/?p=5653
<img width="1076" alt="amp-validation" src="https://user-images.githubusercontent.com/4063887/70651495-8e458380-1c16-11ea-90b8-8d4cdfbc69b9.png">

On clicking 'Validate,' this error appears:
<img width="1002" alt="validation-eher" src="https://user-images.githubusercontent.com/4063887/70651212-06f81000-1c16-11ea-931e-9bc84c45fa45.png">

It might be hard to diagnose that error without the error log, but the Gallery block appears as a carousel in my local when 'Display as carousel' is checked:

<img width="854" alt="gallery-block-elc" src="https://user-images.githubusercontent.com/4063887/70651438-6ce49780-1c16-11ea-8c38-447a06ffb4d9.png">

**Comment by kienstra** [comment]
Also, sometimes there's a lightbox for the Gallery block, even with 'Display as carousel' and 'Add lightbox effect' unselected.

But it looks like this is from 'auto-lightboxing':

https://github.com/ampproject/amphtml/blob/e33dd10533805655835c196389e4b5ea91f080b1/spec/auto-lightbox.md

**Comment by kienstra** [comment]
If it's alright, I'm going to merge this to 'Done.' Feel free to reply if there are more points to bring up.

**Comment by westonruter** [comment]
> It might be hard to diagnose that error without the error log,

You can see the error at the bottom of the page:

```
Fatal error: Uncaught Error: Class 'Amp\AmpWP\Component\DOMElementList' not found in /srv/bindings/842e702ecb134f309437791560d11313/code/wp-content/plugins/amp/includes/sanitizers/class-amp-gallery-block-sanitizer.php:109
Stack trace:
#0 /srv/bindings/842e702ecb134f309437791560d11313/code/wp-content/plugins/amp/includes/templates/class-amp-content-sanitizer.php(117): AMP_Gallery_Block_Sanitizer->sanitize()
#1 /srv/bindings/842e702ecb134f309437791560d11313/code/wp-content/plugins/amp/includes/class-amp-theme-support.php(2315): AMP_Content_Sanitizer::sanitize_document(Object(DOMDocument), Array, Array)
#2 /srv/bindings/842e702ecb134f309437791560d11313/code/wp-content/plugins/amp/includes/class-amp-theme-support.php(2003): AMP_Theme_Support::prepare_response('<!DOCTYPE html>...')
#3 [internal function]: AMP_Theme_Support::finish_output_buffering('<!DOCTYPE html>...', 9)
#4 /srv/bindings/842e702ecb134f309437791560d11313/code/wp-includes/functions.php(4469): ob_end_flush()
#5 /srv/bindings/842e702ecb134f309437791560d11313/code in /srv/bindings/842e702ecb134f309437791560d11313/code/wp-content/plugins/amp/includes/sanitizers/class-amp-gallery-block-sanitizer.php on line 109
```

**Comment by westonruter** [comment]
Fixed by #3921.
